dmake/dbug/malloc/malloc.3
.TH MALLOC 3 "" "" "1.0"
.ds ]T 
.\"/*
.\" * (c) Copyright 1990 Conor P. Cahill (uunet!virtech!cpcahil).  
.\" * You may copy, distribute, and use this software as long as this
.\" * copyright statement is not removed.
.\" */
.\" $Id: malloc.3,v 1.1.1.1 2000-09-22 15:33:26 hr Exp $
.SH NAME
malloc \t- debugging malloc library
.SH SYNOPSIS
.ft B
.nf
#include <malloc.h>

char	* calloc(nelem,elsize);
void	  free(ptr);
char	* malloc(size);
int	  malloc_chain_check(flag);
void	  malloc_dump(fd);
int	  mallopt(cmd,val)
char	* realloc(ptr,size);

int			  cmd,fd,flag;
unsigned		  elsize,nelem,size;
char			* ptr;
union malloptarg	  val;

.fi
.ft R
.SH DESCRIPTION
This malloc library is a replacement for the standard library to be used
during software development/debugging.  See the standard malloc(3) pages
for more information on the use of the following functions:
.nf
.in +.5i
calloc(), free(), malloc(), realloc()
.in -.5i
.fi
.sp
This library differs from the standard malloc library in the
following ways:
.P
1. Each malloc segment contains a magic number so that free can 
verify that the pointer passed points to a valid malloc segment.
.P
2. Each malloc segment is filled with a non-zero pattern so that code that
depends upon malloc segments being null will fail.
.P
3. The size of each segment will be at least 1 byte larger than requested
and the extra bytes will be filled with a non-zero pattern.  When free is
called, it will verify that you did not go beyond the number of bytes 
you asked for.
.P
4. When a segment is freed, it will be filled with a different non-zero pattern
to ensure that the program doesn't depend upon the use of already freed data.
.P
5. Whenever any of the string or memory functions (str*, b*, mem*) are 
called with a pointer that is within the malloc arena,  the operation is
checked to verify that it does not overrun the malloced segment.  A failure
of this check is considered a "warning level error" (described later) and
is handled accordingly.
.P
7. Run time checking can include verification of the malloc chain at each
and every call to one of the malloc functions or manually by calling the
malloc_chain_check function.
.P
6. When a problem is found, the action taken is specified at runtime by
environment variables or at compile time by the use of the mallopt()
function.
.P
There are two arbitrary levels of errors, warning and fatal, that this
library will detect.  They are broken down as follows:
.P
.nf
.in +.25i
Warning messages include:
.sp
.in +.5i
.ti -.25i
Calling free with a bad pointer
.br
.ti -.25i
Calling a bstring/string/memory (3) function which will go beyond
the end of a malloc block. Note that the library function is
not modified to refuse the operation.
.sp
.in -.5i
Fatal errors are:
.in +.5i
.ti -.25i
Detectable corruption to the malloc chain.
.in -.5i
.in -.25i
.P
The error handling for each level (warning or fatal) are specified using
environment variables or mallopt().  The coding for the error handling is
as follows:
.sp
.nf
.in +.5i
.ti -.25i
  0 - continue operations
.ti -.25i
  1 - drop core and exit
.ti -.25i
  2 - just exit
.ti -.25i
  3 - drop core, but continue executing.  Core files will
be placed into core.[PID].[counter] i.e: core.00123.001
.ti -.25i
128 - dump malloc chain and continue
.ti -.25i
129 - dump malloc chain, dump core, and exit
.ti -.25i
130 - dump malloc chain, exit
.ti -.25i
131 - dump malloc chain, dump core, continue processing
.in -.5i
.P
In addition error messages can be placed into an error file.
.P
\fBmalloc_opt\fP() is used to set the malloc debugging options. The
following options can be set:
.br
.sp
.in +.5i
MALLOC_WARN - set the error handling for warning level errors.  \fBval.i\fP is
an integer that can contain any one of the following values:
.sp
.in +.5i
M_HANDLE_IGNORE - ignore error
.br
M_HANDLE_ABORT - drop core and exit
.br
M_HANDLE_EXIT - just exit (no core drop)
.br
M_HANDLE_CORE - drop core, but keep on going
.br
.in -.5i
.sp
In addition, M_HANDLE_DUMP may be or'd in to cause a dump of the current
malloc chain.
.br
.sp
MALLOC_FATAL - set the error handling for fatal level errors.  \fBval.i\fP is
equivalent to \fBval.i\fP for MALLOC_WARN.
.br
.sp
MALLOC_ERRFILE - set the destination for malloc error messages.  \fBval.str\fP
is a pointer to a character string containing the name of the file to be used
for error messages.
.br
.sp
MALLOC_CKCHAIN - set the malloc chain checking flag.  If \fBval.i\fP is
non-zero, chain checking at every call to malloc is turned on.
.br
.sp
For example, to set up the session to generate a core file for
every malloc warning, to drop core and exit on a malloc fatal, and 
to log all messages to the file "malloc_log" do the following:
.sp
.nf
.in +.5i
#include <malloc.h>
malloc_opt(MALLOC_WARN,131);
malloc_opt(MALLOC_FATAL,1);
malloc_opt(MALLOC_ERRFILE,"malloc_log");
.in -.5i
.fi
.in -.5i
.sp
\fBmalloc_opt\fP() can be used to set/alter the debugging options at any
time.
.P
\fBmalloc_dump\fP() will dump a table of the malloc arena showing all
allocated/freed segments and the first few bytes of data in each segment.
\fBfd\fP is the file descriptor to write the data to.
.P
\fBmalloc_chain_check\fP() will check the status of the malloc arena.
If \fBflag\fP is non-zero, an error found in the chain will cause a 
fatal error.  \fBmalloc_chain_check\fP() returns zero when there are no
problems found in the malloc chain, non-zero otherwise.
.SH "ENVIRONMENT VARIABLES"
Environment variables can be used to control error handling, error logging
and malloc chain checking at run time.  The following environment variables
are used:
.P
MALLOC_WARN - specifies the error handling for warning errors
.br
MALLOC_FATAL - specifies the error handling for fatal errors
.br
MALLOC_ERRFILE - specifies the error log file for error messages.  
.br
MALLOC_CKCHAIN - if 1, turns on malloc chain checking at every call to any
of the malloc functions.
.P
For example, to set up the session to generate a core file for
every malloc warning, to drop core and exit on a malloc fatal, and 
to log all messages to the file "malloc_log" do the following:
.sp
.nf
.in +.5i
MALLOC_WARN=131
MALLOC_FATAL=1
MALLOC_ERRFILE=malloc_log

export MALLOC_WARN MALLOC_FATAL MALLOC_ERRFILE
.in -.5i
.fi
.SH WARNINGS
This malloc library and it's associated string and memory functions are
much less efficient than the standard functions due in part to the extra
error checking.  You do not want to use this library when generating a
production (i.e. releasable) version of your software.  It should only
be used during development and testing.
.SH SEE ALSO
stat(2)
.SH AUTHOR
Conor P. Cahill
Virtual Technologies Incorporated
.sp
uunet!virtech!cpcahil


ChangeLog
Please see the output of git log.


dmake/dbug/malloc/malloc.c
/*
 * (c) Copyright 1990 Conor P. Cahill (uunet!virtech!cpcahil).
 * You may copy, distribute, and use this software as long as this
 * copyright statement is not removed.
 */
#include <stdio.h>
#include <fcntl.h>
#include "malloc.h"
#include "tostring.h"

/*
 * Function:    malloc()
 *
 * Purpose: memory allocator
 *
 * Arguments:   size    - size of data area needed
 *
 * Returns: pointer to allocated area, or NULL if unable
 *      to allocate addtional data.
 *
 * Narrative:
 *
 */
#ifndef lint
static
char rcs_hdr[] = "$Id: malloc.c,v 1.2 2006-07-25 10:08:36 rt Exp $";
#endif

extern int    malloc_checking;
char        * malloc_data_start;
char        * malloc_data_end;
struct mlist    * malloc_end;
int       malloc_errfd = 2;
int       malloc_errno;
int       malloc_fatal_level = M_HANDLE_CORE;
struct mlist      malloc_start;
int       malloc_warn_level;
void          malloc_memset();

char *
malloc(size)
    unsigned int      size;
{
    char        * func = "malloc";
    char        * getenv();
    void          malloc_fatal();
    void          malloc_init();
    void          malloc_split();
    void          malloc_warning();
    unsigned int      need;
    struct mlist    * oldptr;
    struct mlist    * ptr;
    char        * sbrk();

    /*
     * If this is the first call to malloc...
     */
    if( malloc_data_start == (char *) 0 )
    {
        malloc_init();
    }

    /*
     * If malloc chain checking is on, go do it.
     */
    if( malloc_checking )
    {
        (void) malloc_chain_check(1);
    }

    /*
     * always make sure there is at least on extra byte in the malloc
     * area so that we can verify that the user does not overrun the
     * data area.
     */
    size++;

    /*
     * Now look for a free area of memory of size bytes...
     */
    oldptr = NULL;
    for(ptr = &malloc_start; ; ptr = ptr->next)
    {
        /*
         * Since the malloc chain is a forward only chain, any
         * pointer that we get should always be positioned in
         * memory following the previous pointer.  If this is not
         * so, we must have a corrupted chain.
         */
        if( ptr )
        {
            if( ptr<oldptr )
            {
                malloc_errno = M_CODE_CHAIN_BROKE;
                malloc_fatal(func);
                return(NULL);
            }
            oldptr = ptr;
        }
        else if( oldptr != malloc_end )
        {
            /*
             * This should never happen.  If it does, then
             * we got a real problem.
             */
            malloc_errno = M_CODE_NO_END;
            malloc_fatal(func);
            return(NULL);
        }


        /*
         * if this element is already in use...
         */
        if( ptr && ((ptr->flag & M_INUSE) != 0) )
        {
            continue;
        }

        /*
         * if there isn't room for this block..
         */
        if( ptr && (ptr->s.size < size) )
        {
            continue;
        }

        /*
         * If ptr is null, we have run out of memory and must sbrk more
         */
        if( ptr == NULL )
        {
            need = (size + M_SIZE) * (size > 10*1024 ? 1:2);
            if( need < M_BLOCKSIZE )
            {
                need = M_BLOCKSIZE;
            }
            else if( need & (M_BLOCKSIZE-1) )
            {
                need &= ~(M_BLOCKSIZE-1);
                need += M_BLOCKSIZE;
            }
            ptr = (struct mlist *) sbrk((int)need);
            if( ptr == (struct mlist *) -1 )
            {
                malloc_errno = M_CODE_NOMORE_MEM;
                malloc_fatal(func);
            }
            malloc_data_end = sbrk((int)0);

            ptr->prev   = oldptr;
            ptr->next   = (struct mlist *) 0;
            ptr->s.size = need - M_SIZE;
            ptr->flag  = M_MAGIC;

            oldptr->next = ptr;
            malloc_end = ptr;


        } /* if( ptr ==... */

        /*
          * Now ptr points to a memory location that can store
         * this data, so lets go to work.
         */

        ptr->r_size = size;     /* save requested size  */
        ptr->flag |= M_INUSE;

        /*
          * split off unneeded data area in this block, if possible...
         */
        malloc_split(ptr);

        /*
         * re-adjust the requested size so that it is what the user
         * actually requested...
         */

        ptr->r_size--;

        /*
         * just to make sure that noone is misusing malloced
          * memory without initializing it, lets set it to
         * all '\01's.  We call local_memset() because memset()
         * may be checking for malloc'd ptrs and this isn't
         * a malloc'd ptr yet.
         */
        malloc_memset(ptr->data,M_FILL,(int)ptr->s.size);

        return( ptr->data);

    } /* for(... */

} /* malloc(... */

/*
 * Function:    malloc_split()
 *
 * Purpose: to split a malloc segment if there is enough room at the
 *      end of the segment that isn't being used
 *
 * Arguments:   ptr - pointer to segment to split
 *
 * Returns: nothing of any use.
 *
 * Narrative:
 *      get the needed size of the module
 *      round the size up to appropriat boundry
 *      calculate amount of left over space
 *      if there is enough left over space
 *          create new malloc block out of remainder
 *          if next block is free
 *          join the two blocks together
 *          fill new empty block with free space filler
 *          re-adjust pointers and size of current malloc block
 *
 *
 *
 * Mod History:
 *   90/01/27   cpcahil     Initial revision.
 */
void
malloc_split(ptr)
    struct mlist        * ptr;
{
    extern struct mlist * malloc_end;
    void              malloc_join();
    int           rest;
    int           size;
    struct mlist        * tptr;

    size = ptr->r_size;

    /*
     * roundup size to the appropriate boundry
     */

    M_ROUNDUP(size);

    /*
     * figure out how much room is left in the array.
     * if there is enough room, create a new mlist
     *     structure there.
     */

    if( ptr->s.size > size )
    {
        rest = ptr->s.size - size;
    }
    else
    {
        rest = 0;
    }

    if( rest > (M_SIZE+M_RND) )
    {
        tptr = (struct mlist *) (ptr->data+size);
        tptr->prev = ptr;
        tptr->next = ptr->next;
        tptr->flag = M_MAGIC;
        tptr->s.size = rest - M_SIZE;

        /*
         * If possible, join this segment with the next one
         */

        malloc_join(tptr, tptr->next,0,0);

        if( tptr->next )
        {
            tptr->next->prev = tptr;
        }

        malloc_memset(tptr->data,M_FREE_FILL, (int)tptr->s.size);

        ptr->next = tptr;
        ptr->s.size = size;

        if( malloc_end == ptr )
        {
            malloc_end = tptr;
        }
    }

} /* malloc_split(... */

/*
 * Function:    malloc_join()
 *
 * Purpose: to join two malloc segments together (if possible)
 *
 * Arguments:   ptr - pointer to segment to join to.
 *      nextptr - pointer to next segment to join to ptr.
 *
 * Returns: nothing of any values.
 *
 * Narrative:
 *
 * Mod History:
 *   90/01/27   cpcahil     Initial revision.
 */
void
malloc_join(ptr,nextptr, inuse_override, fill_flag)
    struct mlist    * ptr;
    struct mlist    * nextptr;
    int       inuse_override;
    int       fill_flag;
{
    unsigned int      newsize;

    if(     ptr     && ! (inuse_override || (ptr->flag & M_INUSE)) &&
        nextptr && ! (nextptr->flag & M_INUSE) &&
        ((ptr->data+ptr->s.size) == (char *) nextptr) )
    {
        if( malloc_end == nextptr )
        {
            malloc_end = ptr;
        }
        ptr->next = nextptr->next;
        newsize = nextptr->s.size + M_SIZE;

        /*
         * if we are to fill and this segment is in use,
         *   fill in with M_FILL newly added space...
          */

        if(fill_flag && (ptr->flag & M_INUSE) )
        {
            malloc_memset(ptr->data+ptr->s.size,
                      M_FILL, (int)(nextptr->s.size + M_SIZE));
        }

        ptr->s.size += newsize;
        if( ptr->next )
        {
            ptr->next->prev = ptr;
        }
    }

} /* malloc_join(... */


/*
 * The following mess is just to ensure that the versions of these functions in
 * the current library are included (to make sure that we don't accidentaly get
 * the libc versions. (This is the lazy man's -u ld directive)
 */

void free();
int strcmp();
int memcmp();
char    * realloc();

void        (*malloc_void_funcs[])() =
{
    free,
};

int     (*malloc_int_funcs[])() =
{
    strcmp,
    memcmp,
};

char        * (*malloc_char_star_funcs[])() =
{
    realloc,
};

/*
 * This is malloc's own memset which is used without checking the parameters.
 */

void
malloc_memset(ptr,byte,len)
    char        * ptr;
    char          byte;
    int       len;
{

    while(len-- > 0)
    {
        *ptr++ = byte;
    }

} /* malloc_memset(... */

/*
 * Function:    malloc_fatal()
 *
 * Purpose: to display fatal error message and take approrpriate action
 *
 * Arguments:   funcname - name of function calling this routine
 *
 * Returns: nothing of any value
 *
 * Narrative:
 *
 * Notes:   This routine does not make use of any libc functions to build
 *      and/or disply the error message.  This is due to the fact that
 *      we are probably at a point where malloc is having a real problem
 *      and we don't want to call any function that may use malloc.
 */
void
malloc_fatal(funcname)
    char        * funcname;
{
    char          errbuf[128];
    void          exit();
    void          malloc_err_handler();
    extern char * malloc_err_strings[];
    extern int    malloc_errno;
    extern int    malloc_fatal_level;
    char        * s;
    char        * t;

    s = errbuf;
    t = "Fatal error: ";
    while( *s = *t++)
    {
        s++;
    }
    t = funcname;
    while( *s = *t++)
    {
        s++;
    }

    t = "(): ";
    while( *s = *t++)
    {
        s++;
    }

    t = malloc_err_strings[malloc_errno];
    while( *s = *t++)
    {
        s++;
    }

    *(s++) = '\n';

    if( write(malloc_errfd,errbuf,(unsigned)(s-errbuf)) != (s-errbuf))
    {
        (void) write(2,"I/O error to error file\n",(unsigned)24);
        exit(110);
    }
    malloc_err_handler(malloc_fatal_level);

} /* malloc_fatal(... */

/*
 * Function:    malloc_warning()
 *
 * Purpose: to display warning error message and take approrpriate action
 *
 * Arguments:   funcname - name of function calling this routine
 *
 * Returns: nothing of any value
 *
 * Narrative:
 *
 * Notes:   This routine does not make use of any libc functions to build
 *      and/or disply the error message.  This is due to the fact that
 *      we are probably at a point where malloc is having a real problem
 *      and we don't want to call any function that may use malloc.
 */
void
malloc_warning(funcname)
    char        * funcname;
{
    char          errbuf[128];
    void          exit();
    void          malloc_err_handler();
    extern char * malloc_err_strings[];
    extern int    malloc_errno;
    extern int    malloc_warn_level;
    char        * s;
    char        * t;

    s = errbuf;
    t = "Warning: ";
    while( *s = *t++)
    {
        s++;
    }
    t = funcname;
    while( *s = *t++)
    {
        s++;
    }

    t = "(): ";
    while( *s = *t++)
    {
        s++;
    }

    t = malloc_err_strings[malloc_errno];
    while( *s = *t++)
    {
        s++;
    }

    *(s++) = '\n';

    if( write(malloc_errfd,errbuf,(unsigned)(s-errbuf)) != (s-errbuf))
    {
        (void) write(2,"I/O error to error file\n",(unsigned)24);
        exit(110);
    }

    malloc_err_handler(malloc_warn_level);

} /* malloc_warning(... */

/*
 * Function:    malloc_err_handler()
 *
 * Purpose: to take the appropriate action for warning and/or fatal
 *      error conditions.
 *
 * Arguments:   level - error handling level
 *
 * Returns: nothing of any value
 *
 * Narrative:
 *
 * Notes:   This routine does not make use of any libc functions to build
 *      and/or disply the error message.  This is due to the fact that
 *      we are probably at a point where malloc is having a real problem
 *      and we don't want to call any function that may use malloc.
 */
void
malloc_err_handler(level)
{
    void          exit();
    void          malloc_dump();
    extern int    malloc_errfd;

    if( level & M_HANDLE_DUMP )
    {
        malloc_dump(malloc_errfd);
    }

    switch( level & ~M_HANDLE_DUMP )
    {
        /*
         * If we are to drop a core file and exit
         */
        case M_HANDLE_ABORT:
            (void) abort();
            break;

        /*
         * If we are to exit..
         */
        case M_HANDLE_EXIT:
            exit(200);
            break;

#ifndef __MSDOS__
        /*
         * If we are to dump a core, but keep going on our merry way
         */
        case M_HANDLE_CORE:
            {
                int   pid;

                /*
                 * fork so child can abort (and dump core)
                 */
                if( (pid = fork()) == 0 )
                {
                    (void) write(2,"Child dumping core\n",
                            (unsigned)9);
                    (void) abort();
                }

                /*
                  * wait for child to finish dumping core
                 */
                while( wait((int *)0) != pid)
                {
                }

                /*
                 * Move core file to core.pid.cnt so
                 * multiple cores don't overwrite each
                 * other.
                 */
                if( access("core",0) == 0 )
                {
                    static int    corecnt;
                    char          filenam[32];
                    filenam[0] = 'c';
                    filenam[1] = 'o';
                    filenam[2] = 'r';
                    filenam[3] = 'e';
                    filenam[4] = '.';
                    (void)tostring(filenam+5,getpid(),
                        5, B_DEC, '0');
                    filenam[10] = '.';
                    (void)tostring(filenam+11,corecnt++,
                        3, B_DEC, '0');
                    filenam[14] = '\0';
                    (void) unlink(filenam);
                    if( link("core",filenam) == 0)
                    {
                        (void) unlink("core");
                    }
                }
            }
#endif


        /*
         * If we are to just ignore the error and keep on processing
         */
        case M_HANDLE_IGNORE:
            break;

    } /* switch(... */

} /* malloc_err_handler(... */

